R/monitor.R

```diff
@@ -0,0 +1,151 @@
+#' Chicago <- Chicago %>% select(ridership, date, one_of(stations))
+#' training_data <- Chicago %>% filter(date < "2009-01-01")
+#' testing_data <- Chicago %>% filter(date >= "2009-01-01", date < "2011-01-01")
+#' monitoring <- Chicago %>% filter(date >= "2011-01-01", date < "2012-12-31")
+#' lm_fit <- linear_reg() %>% fit(ridership ~ ., data = training_data)
+#'
+#' library(pins)
+#' b <- board_temp()
+#'
+#' ## before starting monitoring, initiate the metrics and pin
+#' ## (for example, with the testing data):
+#' original_metrics <-
+#'     augment(lm_fit, new_data = testing_data) %>%
+#'     vetiver_compute_metrics(date, "week", ridership, .pred, .every = 4L) %>%
+#'     vetiver_pin_metrics(date, b, "lm_fit_metrics", initiate = TRUE)
+#'
+#' ## to continue monitoring with new data, compute metrics and update pin:
+#' new_metrics <-
+#'     augment(lm_fit, new_data = monitoring) %>%
+#'     vetiver_compute_metrics(date, "week", ridership, .pred, .every = 4L) %>%
+#'     vetiver_pin_metrics(date, b, "lm_fit_metrics")
+#'
+#' library(ggplot2)
+#' vetiver_plot_metrics(new_metrics, date) +
+#'     scale_size(range = c(2, 4))
+#'
+#' @export
+vetiver_compute_metrics <- function(data,
+                                    date_var,
+                                    .period,
+                                    truth, estimate, ...,
+                                    metric_set = yardstick::metrics,
+                                    .every = 1L,
+                                    .origin = NULL,
+                                    .before = 0L,
+                                    .after = 0L,
+                                    .complete = FALSE) {
+
+    rlang::check_installed("slider")
+    metrics_dots <- list2(...)
+    date_var <- enquo(date_var)
+    slider::slide_period_dfr(
+        data,
+        .i = data[[quo_name(date_var)]],
+        .period = .period,
+        .f = ~ tibble::tibble(
+            !!date_var := min(.x[[quo_name(date_var)]]),
+            n = nrow(.x),
+            metric_set(.x, {{truth}}, {{estimate}}, !!!metrics_dots)
+        ),
+        .every = .every,
+        .origin = .origin,
+        .before = .before,
+        .after = .after,
+        .complete = .complete
+    )
+
+}
+
+#' @rdname vetiver_compute_metrics
+#' @export
+vetiver_pin_metrics <- function(df_metrics,
+                                date_var,
+                                board,
+                                metrics_pin_name,
+                                initiate = FALSE) {
+    date_var <- quo_name(enquo(date_var))
+
+    if (initiate) {
+        new_metrics <- vec_sort(df_metrics)
+    } else {
+        new_dates <- unique(df_metrics[[date_var]])
+        old_metrics <- pins::pin_read(board, metrics_pin_name)
+        old_metrics <- vec_slice(
+            old_metrics,
+            ! old_metrics[[date_var]] %in% new_dates
+        )
+        new_metrics <- vec_sort(vctrs::vec_rbind(old_metrics, df_metrics))
```

Comment:
Am I reading this right that if you supply a new metric for the _same date_ as something that already exists in the pin, then it will overwrite the old one?

Maybe a note in the docs about this would be good.

It is also possible that some people won't want this behavior, or will want a way to opt out of it. I could see wanting to either:
- Error if I accidentally overlapped something in the pin
- Append it anyways

Reply:
This definitely needed more explanation (and I did add an arg to opt out). Thanks! 👍